### PR TITLE
set the template consumer group id first in trigger

### LINKED ISSUE #3328: Knative triggers trying to lookup random UID consumer group id first instead of using the template consumer group id first
**Describe the bug**
Whenever a new Knative trigger is configured looks like its first trying to create a consumer group with Trigger's UID value instead of using the template config value -- "Knative-Trigger-{Namespace}-{trigger name}". This causing issues when using external apache kafka for Borker-trigger implementation. This is a problem especially when we use Kafka cluster that enforce Consumer Group ACL's and we can't use random generated names.

Status log in trigger in namespace:
Error: message: 'unable to query for existing consumergroup: failed to describe consumer
      groups [8b922e66-ba60-4f55-851e-eda1de3d5c30]: kafka server: The client is not
      authorized to access this group'
    reason: Failed to resolve trigger config
    status: "False"

Log in Kafka cluster:
INFO Principal = User:CN=chrysalis-31270b28-f390-4ea1-a229-ebeadeee1572,C=US is Denied Operation = Describe from host = <> on resource = Group:LITERAL:66309158-8e84-4838-b44f-35f8d2fb7feb for request = FindCoordinator with resourceRefCount = 1 (kafka.authorizer.logger)

 **Expected behavior**
Try the template config value for consumer group id first before switching to the random UID.

 **To Reproduce**
Create/ Apply a trigger to the broker which is configured to a Kafka cluster that forces ACL's for consumer group.

 **Knative release version**
Knative - Eventing : 1.10.1
Knative - Serving : 1.10.2

 **Additional context**
Add any other context about the problem here such as proposed priority

Looks like these are lines are the root cause for the issue,
https://github.com/knative-extensions/eventing-kafka-broker/blob/21d21a67ac59afc0c04c96baa01850a62f2c9bb8/control-plane/pkg/reconciler/trigger/trigger.go#L436C2-L455C64

### PR DESCRIPTION
**Pull request by rajibmitra**
Fixes #3328

### CODE REVIEW

File: control-plane/pkg/reconciler/trigger/trigger.go [hunk 1]
```diff
@@ -437,23 +437,28 @@ func (r *Reconciler) reconcileConsumerGroup(ctx context.Context, broker *eventin
...
+		exists, err := kafka.AreConsumerGroupsPresentAndValid(kafkaClusterAdmin, groupID)
 		if err != nil {
-			return false, fmt.Errorf("unable to query for existing consumergroup: %w", err)
+			return false, fmt.Errorf("error checking consumer group exists: %w", err)
 		}
 
-		// No existing consumer groups, use new naming
-		if !valid {
+		if exists {
+			// Consumer group already exists, use it
```
Comment (matzew) [inline control-plane/pkg/reconciler/trigger/trigger.go:450]: that makes sense to me (still we could perhaps name the var differently?)

File: control-plane/pkg/reconciler/trigger/trigger.go [hunk 2]
```diff
@@ -437,23 +437,28 @@ func (r *Reconciler) reconcileConsumerGroup(ctx context.Context, broker *eventin
...
+			// Generate new group ID
 			groupID, err = r.KafkaFeatureFlags.ExecuteTriggersConsumerGroupTemplate(trigger.ObjectMeta)
 			if err != nil {
 				return false, fmt.Errorf("couldn't generate new consumergroup id: %w", err)
 			}
+
+			trigger.Status.Annotations[kafka.GroupIdAnnotation] = groupID
 		}
 
-		trigger.Status.Annotations[kafka.GroupIdAnnotation] = groupID
```
[the remark this answers is not in the document]
Comment (rajibmitra) [reply control-plane/pkg/reconciler/trigger/trigger.go:456]: makes sense.  @pierDipi

File: control-plane/pkg/reconciler/trigger/trigger.go [hunk 3]
```diff
@@ -437,23 +437,28 @@ func (r *Reconciler) reconcileConsumerGroup(ctx context.Context, broker *eventin
...
 	if !ok {
+
 		// Check if a consumer group exists with the old naming convention
 		groupID = string(trigger.UID)
 
-		valid, err := kafka.AreConsumerGroupsPresentAndValid(kafkaClusterAdmin, groupID)
+		exists, err := kafka.AreConsumerGroupsPresentAndValid(kafkaClusterAdmin, groupID)
 		if err != nil {
-			return false, fmt.Errorf("unable to query for existing consumergroup: %w", err)
+			return false, fmt.Errorf("error checking consumer group exists: %w", err)
```
[the remark this answers is not in the document]
Comment (rajibmitra) [reply control-plane/pkg/reconciler/trigger/trigger.go]: changed it.

File: control-plane/pkg/reconciler/trigger/trigger.go [hunk 4]
```diff
@@ -437,23 +437,28 @@ func (r *Reconciler) reconcileConsumerGroup(ctx context.Context, broker *eventin
 	// Existing Triggers might not yet have this annotation
 	groupID, ok := trigger.Status.Annotations[kafka.GroupIdAnnotation]
 	if !ok {
+
 		// Check if a consumer group exists with the old naming convention
 		groupID = string(trigger.UID)
 
-		valid, err := kafka.AreConsumerGroupsPresentAndValid(kafkaClusterAdmin, groupID)
+		exists, err := kafka.AreConsumerGroupsPresentAndValid(kafkaClusterAdmin, groupID)
```
[the remark this answers is not in the document]
Comment (rajibmitra) [reply control-plane/pkg/reconciler/trigger/trigger.go]: removed this logic.

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Comment by Cali0707** [comment]
Hey @rajibmitra to fix the unit tests you will have to initialize the `KafkaFeatureFlags` field in the `Reconciler` struct here: https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/trigger/trigger_test.go#L2952-L3003

For an example of where that is done, look here: https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/broker/broker_test.go#L2865-L2870

And here: https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/broker/broker_test.go#L2932

**Comment by rajibmitra** [comment]
> Hey @rajibmitra to fix the unit tests you will have to initialize the `KafkaFeatureFlags` field in the `Reconciler` struct here:
> 
> https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/trigger/trigger_test.go#L2952-L3003
> 
> For an example of where that is done, look here:
> 
> https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/broker/broker_test.go#L2865-L2870
> 
> And here:
> 
> https://github.com/knative-extensions/eventing-kafka-broker/blob/df977f2b1fcced3a203d0cd29eff3a9b64711438/control-plane/pkg/reconciler/broker/broker_test.go#L2932

Added kafkaFeatureFlags.

**Comment by rajibmitra** [comment]
can anyone please help resolve the unit test cases, it failing even locally after fixing.

**Review by pierDipi — APPROVED** [review]
Thanks @rajibmitra 

/lgtm
/approve

**Comment by pierDipi** [comment]
/override reconciler-tests-namespaced-broker_eventing-kafka-broker_main

**Pull request merged**